File: packages/javascript/src/middleware/retry-logger.ts
import { EventMetadata, Middleware } from '../types';
import config from '../core/config';

export class RetryLoggerMiddleware implements Middleware {
  async call(
    event: Record<string, unknown>,
    metadata: EventMetadata,
    next: () => Promise<void>
  ): Promise<void> {
    const logger = config.logger;

    if (metadata.deliveries && metadata.deliveries > 1) {
      logger.warn('Retrying event', {
        event_id: metadata.event_id,
        subject: metadata.subject,
        delivery_count: metadata.deliveries,
      });
    }

    await next();
  }
}

export default new RetryLoggerMiddleware();

File: packages/ruby/lib/nats_pubsub/middleware/retry_logger.rb
# frozen_string_literal: true

require_relative '../core/logging'

module NatsPubsub
  module Middleware
    # Middleware that logs retry attempts
    class RetryLogger
      def call(subscriber, _payload, metadata)
        if metadata[:deliveries] && metadata[:deliveries] > 1
          max_deliver = NatsPubsub.config.max_deliver

          NatsPubsub::Logging.warn(
            "Retrying #{metadata[:subject]} with #{subscriber.class.name} " \
            "(attempt #{metadata[:deliveries]}/#{max_deliver})",
            tag: 'NatsPubsub::Middleware::RetryLogger'
          )
        end

        yield
      end
    end
  end
end
